### ISSUE TITLE
Azure Stack: Sign-in broken for guest accounts, and discovery process "intractable"

### ISSUE THREAD

**Issue by keystroke** [Microsoft/AzureStorageExplorer#995]
When adding a custom environment (Azure Stack) the storage explorer application assumes the login account used should be used with the home directory. This is not correct, and will result in the following error:

{
  "message": "\"Unexpected token format: urn:ietf:wg:oauth:2.0:oob?error=access_denied&error_description=AADSTS65005%3a+Invalid+resource.+The+client+has+requested+access+to+a+resource+which+is+not+listed+in+the+requested+permissions+in+the+client%27s+application+registration.+Client+app+ID%3a+04b07795-8ddb-461a-bbee-02f9e1bf7b46.+Resource+value+from+request%3a+https%3a%2f%2fadminmanagement.msazurestack.onmicrosoft.com%2f9f4c9261-e5d2-4e83-bc6c-763f54de5046.+Resource+app+ID%3a+42547758-86d7-45f3-81f8-183181e67712.+List+of+valid+resources+from+app+registration%3a+.%0d%0aTrace+ID%3a+7428aebd-5a29-451c-aae4-611389842b00%0d%0aCorrelation+ID%3a+7ecb73fb-6508-4f0c-abad-d4987af5e641%0d%0aTimestamp%3a+2018-12-11+23%3a36%3a42Z\""
}

Above happens when you use an account which is only a guest in the "home" directory of an Azure Stack. The storage explorer code assumes they can use the common login endpoint and acquire a token for the target ARM application in that directory. This is a permission that has to be manually granted when a directory is onboarded to Azure Stack (we create permission grants for Azure PowerShell and Azure CLI applications to acquire a token against the specific ARM instance, and it looks like Storage Explorer is using the app ID for Azure CLI).

---

Additionally, the discovery process used is extremely inefficient and resulted in over 40+ requests in the background, consuming 100% CPU and only terminating when the access tokens expired, which caused it to restart the process over again. When a custom environment is added, there should not be any enumeration of SPs in the authenticated directory, and instead only the ARM resource as advertised at the metadata endpoint should be used to acquire a token, and then subsequently call the /subscriptions API at that endpoint.

Below image shows how tokens are repeatedly acquired for resources in many different directories and tenant details are queried. This process essentially continues forever consuming maximum resources in the background. While this is occurring, the accounts tab has a spinning icon indicating to the user it is trying to load their subscriptions. To seems to have been developed with accounts that do not belong to many directories, and to directories without many Azure Stack instances registered to them.
![storageexplorercalls](https://user-images.githubusercontent.com/2002281/49842208-7de29a80-fd6f-11e8-93fb-21410ca7a613.png)

**Comment by MRayermannMSFT** [comment]
I'll address each thing 1 by 1:

1. Guest accounts for Azure Stack
Can you expand on this? Or I can try seeing if I understand. We enumerate SPs (applications) looking for Azure Stacks. When we find one, we try to get its endpoints. From there, we get a login endpoint, which we eventually use as the resource for getting tokens for doing things like listing resources. You're saying that if an account is a guest, then it may not be able to use that endpoint, right?

2. Enumerating SPs for custom environments
You're probably right that we could assume that we don't need to enumerate SPs for accounts added via a custom environment. This problem should be fairly easy to solve, though we might need to think about it a bit more/figure out if we need to at least expose the ability.

3. This feature being developed with accounts that don't have a lot of Azure Stack instances
Yes that is true! 😄 That being said, we knew there might be people with lots of Azure Stack instances on their account, but we wanted to wait for feedback that such people really existed before trying to solve the problem. I believe you're the first person to let us know that due to such a situation, loading your subscriptions is taking a reallllllly long time. We'll start thinking about how best to solve the problem. Do you have any initial thoughts on what you might like to see from the user experience standpoint?

**Comment by jinglouMSFT** [comment]
Just to make sure that you have targeted Storage Explorer for Azure Stack by using the Edit->Target Azure Stack APIs menu, correct?

**Comment by keystroke** [comment]
>           Just to make sure that you have targeted Storage Explorer for Azure Stack by using the Edit->Target Azure Stack APIs menu, correct?

Yes I have switched to target. I'm preparing more details about the discovery process, but I'm having trouble aborting it as it seems to be stuck in an infinite loop making the same Graph calls, and when I select "remove account" it looks like it is waiting for the discovery process to complete before I can remove it :(

**Comment by MRayermannMSFT** [comment]
Hey @keystroke , thank you for the thorough replies! You answer about the common endpoint is something that I've been trying to get information on for a long time. I think I agree with most everything you're suggesting/saying, and I definitely want to get the guest tenant situation figured out. It is a scenario the Azure Stack team has wanted us to support, but they never quite explained what we need to do to support it (though you've explained it quite nicely 😄 ). We'll try to get this done as soon as possible, but thank you in advance for your patience as I'm not sure how soon that will be. I may post on this issue in the future with questions, and if you'd like, I can also reach back out once we have something for you to try out.

**Comment by keystroke** [comment]
Any update on this? Storage explorer is completely unusable on my machine even for connecting to Azure, as it blocks at the splash screen and consumes 100% CPU making endless graph queries in the background.
